Kitodo/src/main/java/org/kitodo/production/helper/StringComparator.java

```diff
@@ -0,0 +1,365 @@
+/*
+ * (c) Kitodo. Key to digital objects e. V. <contact@kitodo.org>
+ *
+ * This file is part of the Kitodo project.
+ *
+ * It is licensed under GNU General Public License version 3 or later.
+ *
+ * For the full copyright and license information, please read the
+ * GPL3-License.txt file that was distributed with this source code.
+ */
+
+package org.kitodo.production.helper;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+/**
+ * A String comparator implementation that compares Strings as to their expected
+ * sort order. This especially respects ordering by type (with symbols come
+ * before numbers before upper case letters before lower case letters), sorting
+ * umlauts and German sharp s as to German sort rules, and sorting sequences of
+ * numbers by their numeric value.
+ */
+public class StringComparator implements Comparator<String> {
```

Comment:
Is normal String class comparation not working for German language? I would suspect that UTF-8 contains all letters (ASCII table), also umlauts. This class looks for me as not needed overhead.

Reply:
No. Sorting in German is complicated. The correct sorting of umlauts (Ä = AE, ß = SS) is only a part of it. In addition, symbols must be sorted before numbers, before letters, and it must first be sorted case-insensitive, and only if that is the same, it must be sorted case-sensitive. And above all, there is no sorting function that sorts numbers in their mathematical order if they are part of strings but do not have leading zeros. This class accomplishes both. It does exactly the result you would expect intuitively.

Reply:
1. There is no tests for it - no prove that this class accomplishes anything.

2. Sorting umlauts can be accomplished in very simple way:
`Comparator umlautCollator = Collator.getInstance(Locale.GERMAN);
    public int compare(String o1, String o2) {
        return umlautCollator.compare(o1, o2);
    }`

3. _no sorting function that sorts numbers in their mathematical order_ - so in which order? standard sort just assign it randomly?

Reply:
3. If you have files from `Image1` to `Image102`, standard sort will sort them `Image1`, `Image10`, `Image100`, `Image101`, `Image102`, `Image11`, `Image12`, `Image13`, …

Reply:
I will add a test case. 👍

Reply:
Some elements:

Reply:
I will add a test case. 👍 

FYI. Some elements (correct order, symbol first, numbers in mathematical order, „Dänemark“ between „Dadaismus“ and „dafür“, „Gen“ next to „gen“ and uppercase first, correct image file order): ☆, 8, 9, 10, 11, Dadaismus, Dänemark, dafür, Füssen, Füßen, Fuß, Gen, gen, Image7.jpg, Image008.jpg, Image08.jpg, Image8.jpg, Image9.jpg, Image10.jpg, Image11.jpg, images/

When used as input in Treemap with **default comparator**, produces: 10, 11, 8, 9, Dadaismus, Dänemark, Fuß, Füssen, Füßen, Gen, Image008.jpg, Image08.jpg, Image10.jpg, Image11.jpg, Image7.jpg, Image8.jpg, Image9.jpg, dafür, gen, images/, ☆
→ does not sort symbols before numbers, does not sort numbers in correct order, does not ignore case unless two components are equal if compared case-insensitive, does not treat umlauts, order of image files is a mess

When sorted with `new TreeSet<>(Collator.getInstance(Locale.GERMAN))`: 10, 11, 8, 9, Dadaismus, dafür, Dänemark, Fuß, Füssen, Füßen, gen, Gen, Image008.jpg, Image08.jpg, Image10.jpg, Image11.jpg, Image7.jpg, Image8.jpg, Image9.jpg, images/, ☆
→ does not sort symbols before numbers, does not sort numbers in correct order, puts lowercase before Uppercase if two components are equal if compared case-insensitive, order of image files is a mess

I added this comparator class here especially because to support the correct order of images. But it can also be used elsewhere in the application for sorting lists on the user interface which will produce dictionary-style order.